analyzer/protocol/openvpn/openvpn.spicy

```diff
@@ -0,0 +1,181 @@
+		Opcode::P_CONTROL_SOFT_RESET_V1 -> control_soft_reset_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_V1 -> control_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_ACK_V1 -> ack_v1: AckMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_DATA_V1 -> data_v1: DataMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_HARD_RESET_SERVER_V2 -> control_hard_reset_server_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V2 -> control_hard_reset_client_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_DATA_V2 -> data_v2: DataMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+	};
+};
+
+public type OpenVPNRecordsHMAC = unit {
+	%context=Share;
+	records : OpenVPNRecordHMAC(self.context())[];
+};
+
+public type OpenVPNRecordHMAC = unit(inout share: Share) {
+	message_type: bitfield(8) {
+		opcode: 3..7 &convert=Opcode($$);
+		key_id: 0..2;
+	};
+
+	switch ( self.message_type.opcode ) {
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V1 -> control_hard_reset_client_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_CONTROL_HARD_RESET_SERVER_V1 -> control_hard_reset_server_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_CONTROL_SOFT_RESET_V1 -> control_soft_reset_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_CONTROL_V1 -> control_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_ACK_V1 -> ack_v1: AckMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_DATA_V1 -> data_v1: DataMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_HARD_RESET_SERVER_V2 -> control_hard_reset_server_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V2 -> control_hard_reset_client_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_DATA_V2 -> data_v2: DataMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+	};
+};
+
+public type OpenVPNRecordsTCP = unit {
+	%context=Share;
+	records : OpenVPNRecordTCP(self.context())[];
+};
+
+public type OpenVPNRecordTCP = unit(inout share: Share) {
+	packet_length : uint16;
+
+	message_type: bitfield(8) {
+		opcode: 3..7 &convert=Opcode($$);
+		key_id: 0..2;
+	};
+
+	switch ( self.message_type.opcode ) {
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V1 -> control_hard_reset_client_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share) &size=self.packet_length-1;
```

Comment:
It'd be nice if switch allowed to add a single &size that applied to all fields. I recorded a ticket for that: zeek/spicy#844

Reply:
Nice, I was thinking the same thing when I wrote this.